JavaScript/Angular.md

```diff
@@ -795,6 +795,384 @@ export class AuthorModel extends User {
+
+However, there are some cases where you will have to use `ngOnInit` or `ngOnChanges`.
+
+_Example #4_ - when are `ngOnInit` and `ngOnChanges` necessary?
+
+```ts
+// bad
+@Component(...)
+class PersonDetailsComponent {
+  @Input() public birthDate: Date;
+  public isLegalAge: boolean;
+
+  constructor() {
+    this.isLegalAge = new Date().getYear() - this.birthDate.getYear() > 18;
+    // I know this check is not 100% correct, just keeping it simple.
+  }
+}
+```
+
+If you try this, you will get an exception because `birthDate` will be undefined at component construction time. Input bindings values are available only after `ngOnInit` and `ngOnChanges` are called. To make this work, you can use `ngOnInit`:
+
+```ts
+// bad
+@Component(...)
+class PersonDetailsComponent implements OnInit {
+  @Input() public birthDate: Date;
+  public isLegalAge: boolean;
+
+  ngOnInit() {
+    this.isLegalAge = new Date().getYear() - this.birthDate.getYear() > 18;
+  }
+}
+```
+
+Using `ngOnInit` works and we get no exception (with the assumption that a valid `Date` object is passed down to our component via `birthDate` input).
+
+We still have one problem with this solution. If input binding changes, `isLegalAge` will not be re-assigned. To solve this we simply switch to useing `ngOnChanges`:
+
+```ts
+// good
+@Component(...)
+class PersonDetailsComponent implements OnInit {
+  @Input() public birthDate: Date;
+  public isLegalAge: boolean;
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (changes.birthDate) {
+      this.isLegalAge = this.checkIfLegalAge(changes.birthDate.currentValue);
+    }
+  }
+
+  private checkIfLegalAge(birthDate): boolean {
+    return new Date().getYear() - birthDate.getYear() > 18;
+  }
+}
+```
+
+Since this check is simple and returns a primitive value, using a getter here is also a valid option and it allows us to remove ngOnChanges completely and reduce the amount of code.
+
+```ts
+// even better
+@Component(...)
+class PersonDetailsComponent implements OnInit {
+  @Input() public birthDate: Date;
+
+  public get isLegalAge(): boolean {
+    return new Date().getYear() - this.birthDate.getYear() > 18;
+  }
+}
+```
+
+But be careful when using getters for values which are used in templates because they will be called very often. This is OK only if there is not too much heavy lifting inside getters and if the getter does not return a newly initialized object every time it is called.
+
+_Example #5_ - avoid unnecessary creation of multiple observables and avoid subscriptions
+
+Taking learnings from all previous examples and the fact that we can use services during initial assignment, we can greatly simplify initialization of observables and their usage in templates.
+
+```ts
+// bad
+@Component({
+  template: `
+  <ng-container *ngIf="user">
+    {{ user.name }}
+  </ng-container>
+  `,
+  ...
+})
+class MyComponent {
+  public user: UserModel = this.route.params.pipe(switchMap((params: Params) => {
```

Comment:
`user$`?

Reply:
Actually, no, it should stay `user`, but initial assignment should be removed.